File: java/geometry/ConvexHull.java
package geometry;

import java.util.*;

public class ConvexHull {
    public static Point[] convexHull(Point[] points) {
        Arrays.sort(points, Comparator.<Point>comparingInt(p -> p.x).thenComparingInt(p -> p.y));
        int n = points.length;
        Point[] hull = new Point[n + 1];
        int cnt = 0;
        for (int i = 0; i < 2 * n - 1; i++) {
            int j = i < n ? i : 2 * n - 2 - i;
            while (cnt >= 2 && isNotRightTurn(hull[cnt - 2], hull[cnt - 1], points[j])) --cnt;
            hull[cnt++] = points[j];
        }
        return Arrays.copyOf(hull, cnt - 1);
    }

    static boolean isNotRightTurn(Point a, Point b, Point c) {
        long cross = (long) (a.x - b.x) * (c.y - b.y) - (long) (a.y - b.y) * (c.x - b.x);
        long dot = (long) (a.x - b.x) * (c.x - b.x) + (long) (a.y - b.y) * (c.y - b.y);
        return cross < 0 || cross == 0 && dot <= 0;
    }

    public static class Point {
        public final int x, y;

        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }

        @Override
        public String toString() {
            return "Point{"
                + "x=" + x + ", y=" + y + '}';
        }
    }

    // Usage example
    public static void main(String[] args) {
        Point[] points = {new Point(0, 0), new Point(0, 1), new Point(1, 0), new Point(0, 0)};
        Point[] convexHull = convexHull(points);
        System.out.println(Arrays.toString(convexHull));
    }
}

File: cpp/geometry/convex_hull.cpp
#include <bits/stdc++.h>

using namespace std;

// Convex hull construction in O(n*log(n)): https://cp-algorithms.com/geometry/grahams-scan-convex-hull.html

struct point {
    int x, y;
};

bool isNotRightTurn(const point &a, const point &b, const point &c) {
    long long cross = (long long)(a.x - b.x) * (c.y - b.y) - (long long)(a.y - b.y) * (c.x - b.x);
    long long dot = (long long)(a.x - b.x) * (c.x - b.x) + (long long)(a.y - b.y) * (c.y - b.y);
    return cross < 0 || (cross == 0 && dot <= 0);
}

vector<point> convex_hull(vector<point> points) {
    sort(points.begin(), points.end(), [](auto a, auto b) { return a.x < b.x || (a.x == b.x && a.y < b.y); });
    int n = points.size();
    vector<point> hull;
    for (int i = 0; i < 2 * n - 1; i++) {
        int j = i < n ? i : 2 * n - 2 - i;
        while (hull.size() >= 2 && isNotRightTurn(hull.end()[-2], hull.end()[-1], points[j]))
            hull.pop_back();
        hull.push_back(points[j]);
    }
    hull.pop_back();
    return hull;
}

// usage example
int main() {
    vector<point> hull1 = convex_hull({{0, 0}, {3, 0}, {0, 3}, {1, 1}});
    cout << (3 == hull1.size()) << endl;

    vector<point> hull2 = convex_hull({{0, 0}, {0, 0}});
    cout << (1 == hull2.size()) << endl;
}
